File: supabase/supabase_client.js
/**
 * Supabase JavaScript Client Example
 *
 * This module demonstrates how to configure and use the Supabase JavaScript client
 * for database operations, authentication, and storage.
 *
 * Requirements:
 *   npm install @supabase/supabase-js
 *
 * Usage (Node.js):
 *   import { getClient, fetchBooks, signInWithGoogle } from './supabase_client.js';
 *
 *   const client = getClient();
 *   const { data: books } = await fetchBooks({ limit: 10 });
 *
 * Usage (Browser/React):
 *   import { supabase, signInWithGoogle } from './supabase_client.js';
 *
 *   const handleLogin = async () => {
 *     const { error } = await signInWithGoogle();
 *     if (error) console.error(error);
 *   };
 *
 * Security Notes:
 *   - SUPABASE_ANON_KEY: Safe for client-side, respects RLS
 *   - SUPABASE_SERVICE_ROLE_KEY: Server-side only, bypasses RLS
 *   - NEVER expose the service role key in browser/frontend code!
 */

// For Node.js environments
// const { createClient } = require('@supabase/supabase-js');

// For ES modules / browser
import { createClient } from '@supabase/supabase-js';

// ============================================================================
// Configuration
// ============================================================================

// For browser (React/Vue/etc.), use environment variables with REACT_APP_ or VITE_ prefix
const SUPABASE_URL = process.env.SUPABASE_URL ||
    process.env.REACT_APP_SUPABASE_URL ||
    process.env.VITE_SUPABASE_URL;

const SUPABASE_ANON_KEY = process.env.SUPABASE_ANON_KEY ||
    process.env.REACT_APP_SUPABASE_ANON_KEY ||
    process.env.VITE_SUPABASE_ANON_KEY;

// Validate configuration
if (!SUPABASE_URL || !SUPABASE_ANON_KEY) {
    throw new Error(
        'Supabase configuration missing. Set SUPABASE_URL and SUPABASE_ANON_KEY ' +
        'in your environment variables.'
    );
}

// ============================================================================
// Client Initialization
// ============================================================================

/**
 * Supabase client instance.
 *
 * This client uses the anon key and respects Row Level Security (RLS).
 * Safe to use in browser/frontend code.
 */
export const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY, {
    auth: {
        persistSession: true,
        autoRefreshToken: true,
        detectSessionInUrl: true
    }
});

/**
 * Get Supabase client (alternative getter function).
 *
 * @returns {SupabaseClient} The Supabase client instance
 */
export function getClient() {
    return supabase;
}

// ============================================================================
// Authentication Functions
// ============================================================================

/**
 * Sign in with Google OAuth.
 *
 * This redirects the user to Google's OAuth page and back to your app.
 *
 * @param {string} redirectTo - URL to redirect to after sign-in
 * @returns {Promise<{data, error}>} Auth response
 *
 * @example
 * const { error } = await signInWithGoogle('https://myapp.com/dashboard');
 */
export async function signInWithGoogle(redirectTo) {
    const { data, error } = await supabase.auth.signInWithOAuth({
        provider: 'google',
        options: {
            redirectTo: redirectTo || window.location.origin
        }
    });
    return { data, error };
}

/**
 * Sign in with email and password.
 *
 * @param {string} email - User's email
 * @param {string} password - User's password
 * @returns {Promise<{data, error}>} Auth response
 *
 * @example
 * const { data, error } = await signInWithEmail('user@example.com', 'password123');
 */
export async function signInWithEmail(email, password) {
    const { data, error } = await supabase.auth.signInWithPassword({
        email,
        password
    });
    return { data, error };
}

/**
 * Sign up with email and password.
 *
 * @param {string} email - User's email
 * @param {string} password - User's password
 * @param {Object} metadata - Optional user metadata
 * @returns {Promise<{data, error}>} Auth response
 */
export async function signUpWithEmail(email, password, metadata = {}) {
    const { data, error } = await supabase.auth.signUp({
        email,
        password,
        options: {
            data: metadata
        }
    });
    return { data, error };
}

/**
 * Sign out the current user.
 *
 * @returns {Promise<{error}>} Sign out response
 */
export async function signOut() {
    const { error } = await supabase.auth.signOut();
    return { error };
}

/**
 * Get the current authenticated user.
 *
 * @returns {Promise<{data, error}>} User data or null
 */
export async function getCurrentUser() {
    const { data: { user }, error } = await supabase.auth.getUser();
    return { user, error };
}

/**
 * Listen for authentication state changes.
 *
 * @param {Function} callback - Function to call on auth state change
 * @returns {Function} Unsubscribe function
 *
 * @example
 * const unsubscribe = onAuthStateChange((event, session) => {
 *   if (event === 'SIGNED_IN') {
 *     console.log('User signed in:', session.user);
 *   }
 * });
 */
export function onAuthStateChange(callback) {
    const { data: { subscription } } = supabase.auth.onAuthStateChange(callback);
    return () => subscription.unsubscribe();
}

// ============================================================================
// Database Operations
// ============================================================================

/**
 * Fetch books from the database.
 *
 * @param {Object} options - Query options
 * @param {number} options.limit - Maximum number of books (default: 10)
 * @param {number} options.offset - Number of books to skip (default: 0)
 * @param {string} options.type - Filter by book type
 * @param {string} options.status - Filter by status
 * @returns {Promise<{data, error}>} Books data
 *
 * @example
 * const { data: books, error } = await fetchBooks({
 *   limit: 20,
 *   type: 'manga',
 *   status: 'ongoing'
 * });
 */
export async function fetchBooks({ limit = 10, offset = 0, type, status } = {}) {
    let query = supabase
        .from('books')
        .select(`
            *,
            authors (id, name),
            publishers (id, name)
        `)
        .order('created_at', { ascending: false })
        .range(offset, offset + limit - 1);

    if (type) {
        query = query.eq('type', type);
    }

    if (status) {
        query = query.eq('status', status);
    }

    const { data, error } = await query;
    return { data, error };
}

/**
 * Get a single book by ID.
 *
 * @param {string} bookId - UUID of the book
 * @returns {Promise<{data, error}>} Book data
 */
export async function getBookById(bookId) {
    const { data, error } = await supabase
        .from('books')
        .select(`
            *,
            authors (id, name, biography),
            publishers (id, name),
            book_tags (
                tags (id, name)
            )
        `)
        .eq('id', bookId)
        .single();

    return { data, error };
}

/**
 * Create a review for a book.
 *
 * @param {Object} review - Review data
 * @param {string} review.bookId - UUID of the book
 * @param {number} review.rating - Rating from 1-5
 * @param {string} review.content - Review content
 * @param {string} review.title - Optional review title
 * @param {boolean} review.isSpoiler - Whether review contains spoilers
 * @returns {Promise<{data, error}>} Created review
 */
export async function createReview({ bookId, rating, content, title = null, isSpoiler = false }) {
    const { data: { user } } = await supabase.auth.getUser();

    if (!user) {
        return { data: null, error: { message: 'User must be authenticated to create reviews' } };
    }

    const { data, error } = await supabase
        .from('reviews')
        .insert({
            user_id: user.id,
            book_id: bookId,
            rating,
            content,
            title,
            is_spoiler: isSpoiler
        })
        .select()
        .single();

    return { data, error };
}

/**
 * Get reviews for a book.
 *
 * @param {string} bookId - UUID of the book
 * @param {Object} options - Query options
 * @returns {Promise<{data, error}>} Reviews data
 */
export async function getBookReviews(bookId, { limit = 10, offset = 0 } = {}) {
    const { data, error } = await supabase
        .from('reviews')
        .select(`
            *,
            users (id, username, display_name, avatar_url)
        `)
        .eq('book_id', bookId)
        .eq('is_approved', true)
        .order('created_at', { ascending: false })
        .range(offset, offset + limit - 1);

    return { data, error };
}

/**
 * Add a book to favorites.
 *
 * @param {string} bookId - UUID of the book
 * @param {string} listName - List type (favorites, reading, completed, plan_to_read, dropped)
 * @returns {Promise<{data, error}>} Created favorite
 */
export async function addToFavorites(bookId, listName = 'favorites') {
    const { data: { user } } = await supabase.auth.getUser();

    if (!user) {
        return { data: null, error: { message: 'User must be authenticated' } };
    }

    const { data, error } = await supabase
        .from('favorites')
        .insert({
            user_id: user.id,
            book_id: bookId,
            list_name: listName
        })
        .select()
        .single();

    return { data, error };
}

/**
 * Get user's favorites.
 *
 * @param {string} listName - Optional filter by list type
 * @returns {Promise<{data, error}>} Favorites with book details
 */
export async function getFavorites(listName = null) {
    const { data: { user } } = await supabase.auth.getUser();

    if (!user) {
        return { data: null, error: { message: 'User must be authenticated' } };
    }

    let query = supabase
        .from('favorites')
        .select(`
            *,
            books (id, title, cover_image_url, type, status, average_rating)
        `)
        .eq('user_id', user.id)
        .order('created_at', { ascending: false });

    if (listName) {
        query = query.eq('list_name', listName);
    }

    const { data, error } = await query;
    return { data, error };
}

/**
 * Remove a book from favorites.
 *
 * @param {string} bookId - UUID of the book
 * @param {string} listName - Optional list name filter
 * @returns {Promise<{error}>} Delete result
 */
export async function removeFromFavorites(bookId, listName = null) {
    const { data: { user } } = await supabase.auth.getUser();

    if (!user) {
        return { error: { message: 'User must be authenticated' } };
    }

    let query = supabase
        .from('favorites')
        .delete()
        .eq('user_id', user.id)
        .eq('book_id', bookId);

    if (listName) {
        query = query.eq('list_name', listName);
    }

    const { error } = await query;
    return { error };
}

// ============================================================================
// Storage Functions
// ============================================================================

/**
 * Upload a file to Supabase Storage.
 *
 * @param {string} bucket - Storage bucket name
 * @param {string} path - File path in the bucket
 * @param {File|Blob} file - File to upload
 * @returns {Promise<{data, error}>} Upload result
 */
export async function uploadFile(bucket, path, file) {
    const { data, error } = await supabase.storage
        .from(bucket)
        .upload(path, file, {
            cacheControl: '3600',
            upsert: false
        });

    return { data, error };
}

/**
 * Get public URL for a file.
 *
 * @param {string} bucket - Storage bucket name
 * @param {string} path - File path in the bucket
 * @returns {string} Public URL
 */
export function getPublicUrl(bucket, path) {
    const { data } = supabase.storage.from(bucket).getPublicUrl(path);
    return data.publicUrl;
}

// ============================================================================
// Real-time Subscriptions
// ============================================================================

/**
 * Subscribe to real-time changes on a table.
 *
 * @param {string} table - Table name to subscribe to
 * @param {Function} callback - Function to call on changes
 * @param {Object} options - Subscription options
 * @returns {Object} Subscription channel (call .unsubscribe() to stop)
 *
 * @example
 * const channel = subscribeToTable('reviews', (payload) => {
 *   console.log('Review changed:', payload);
 * }, { filter: 'book_id=eq.some-uuid' });
 *
 * // Later, to unsubscribe:
 * channel.unsubscribe();
 */
export function subscribeToTable(table, callback, { filter, event = '*' } = {}) {
    let channel = supabase
        .channel(`public:${table}`)
        .on(
            'postgres_changes',
            {
                event,
                schema: 'public',
                table,
                filter
            },
            callback
        )
        .subscribe();

    return channel;
}

// ============================================================================
// Export default client for convenience
// ============================================================================

export default supabase;

File: supabase/supabase_client.py
"""
Supabase Python Client Example

This module demonstrates how to configure and use the Supabase Python client
for database operations, authentication, and storage.

Requirements:
    pip install supabase python-dotenv

Usage:
    from supabase.supabase_client import get_client, fetch_books, create_review
    
    # Get the client
    client = get_client()
    
    # Use helper functions
    books = fetch_books(limit=10)
    review = create_review(user_id, book_id, rating=5, content="Great book!")

Security Notes:
    - SUPABASE_ANON_KEY: Safe for client-side, respects RLS
    - SUPABASE_SERVICE_ROLE_KEY: Server-side only, bypasses RLS
    - Never expose the service role key to frontend code!
"""

import os
from typing import Optional, List, Dict, Any
from functools import lru_cache

# Optional: Load from .env file
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass


def _get_env(name: str, required: bool = True) -> Optional[str]:
    """Get environment variable with optional requirement check."""
    value = os.getenv(name)
    if required and not value:
        raise ValueError(
            f"Missing required environment variable: {name}. "
            "Set it in your .env file or environment."
        )
    return value


@lru_cache(maxsize=1)
def get_client():
    """
    Get Supabase client using the anon key.
    
    This client respects Row Level Security (RLS) policies and is safe
    for user-facing operations.
    
    Returns:
        Supabase client instance
        
    Example:
        client = get_client()
        response = client.table("books").select("*").limit(10).execute()
        books = response.data
    """
    try:
        from supabase import create_client, Client
    except ImportError:
        raise ImportError(
            "The 'supabase' package is required. "
            "Install it with: pip install supabase"
        )
    
    url = _get_env("SUPABASE_URL")
    anon_key = _get_env("SUPABASE_ANON_KEY")
    
    return create_client(url, anon_key)


@lru_cache(maxsize=1)
def get_admin_client():
    """
    Get Supabase client using the service role key.
    
    WARNING: This client bypasses Row Level Security!
    Use only for server-side admin operations.
    
    Returns:
        Supabase admin client instance
    """
    try:
        from supabase import create_client, Client
    except ImportError:
        raise ImportError(
            "The 'supabase' package is required. "
            "Install it with: pip install supabase"
        )
    
    url = _get_env("SUPABASE_URL")
    service_role_key = _get_env("SUPABASE_SERVICE_ROLE_KEY")
    
    return create_client(url, service_role_key)


# ============================================================================
# Example CRUD Operations
# ============================================================================
# These functions demonstrate common patterns. Adapt table names and columns
# to match your actual schema.

def fetch_books(
    limit: int = 10,
    offset: int = 0,
    book_type: Optional[str] = None,
    status: Optional[str] = None
) -> List[Dict[str, Any]]:
    """
    Fetch books from the database with optional filters.
    
    Args:
        limit: Maximum number of books to return
        offset: Number of books to skip (for pagination)
        book_type: Filter by type (manga, novel, etc.)
        status: Filter by status (ongoing, completed, etc.)
        
    Returns:
        List of book dictionaries
        
    Example:
        # Get first 10 manga
        manga = fetch_books(limit=10, book_type="manga")
        
        # Get completed novels
        novels = fetch_books(book_type="novel", status="completed")
    """
    client = get_client()
    
    query = client.table("books").select(
        "*",
        "authors(id, name)",
        "publishers(id, name)"
    )
    
    if book_type:
        query = query.eq("type", book_type)
    
    if status:
        query = query.eq("status", status)
    
    response = query.order("created_at", desc=True).range(offset, offset + limit - 1).execute()
    
    return response.data


def get_book_by_id(book_id: str) -> Optional[Dict[str, Any]]:
    """
    Get a single book by its ID.
    
    Args:
        book_id: UUID of the book
        
    Returns:
        Book dictionary or None if not found
    """
    client = get_client()
    
    response = client.table("books").select(
        "*",
        "authors(id, name, biography)",
        "publishers(id, name)",
        "book_tags(tags(id, name))"
    ).eq("id", book_id).single().execute()
    
    return response.data


def create_review(
    user_id: str,
    book_id: str,
    rating: int,
    content: str,
    title: Optional[str] = None,
    is_spoiler: bool = False
) -> Dict[str, Any]:
    """
    Create a new review for a book.
    
    Args:
        user_id: UUID of the user creating the review
        book_id: UUID of the book being reviewed
        rating: Rating from 1-5
        content: Review text content
        title: Optional review title
        is_spoiler: Whether the review contains spoilers
        
    Returns:
        The created review data
        
    Raises:
        Exception: If the review already exists (user can only review once per book)
    """
    client = get_client()
    
    review_data = {
        "user_id": user_id,
        "book_id": book_id,
        "rating": rating,
        "content": content,
        "title": title,
        "is_spoiler": is_spoiler
    }
    
    response = client.table("reviews").insert(review_data).execute()
    
    return response.data[0] if response.data else None


def update_review(
    review_id: str,
    user_id: str,
    updates: Dict[str, Any]
) -> Optional[Dict[str, Any]]:
    """
    Update an existing review.
    
    Note: RLS ensures users can only update their own reviews.
    
    Args:
        review_id: UUID of the review
        user_id: UUID of the user (for RLS verification)
        updates: Dictionary of fields to update
        
    Returns:
        Updated review data or None
    """
    client = get_client()
    
    # Only allow updating certain fields
    allowed_fields = {"rating", "content", "title", "is_spoiler"}
    safe_updates = {k: v for k, v in updates.items() if k in allowed_fields}
    
    response = client.table("reviews")\
        .update(safe_updates)\
        .eq("id", review_id)\
        .eq("user_id", user_id)\
        .execute()
    
    return response.data[0] if response.data else None


def delete_review(review_id: str, user_id: str) -> bool:
    """
    Delete a review.
    
    Note: RLS ensures users can only delete their own reviews.
    
    Args:
        review_id: UUID of the review
        user_id: UUID of the user (for RLS verification)
        
    Returns:
        True if deleted, False otherwise
    """
    client = get_client()
    
    response = client.table("reviews")\
        .delete()\
        .eq("id", review_id)\
        .eq("user_id", user_id)\
        .execute()
    
    return len(response.data) > 0


def add_to_favorites(
    user_id: str,
    book_id: str,
    list_name: str = "favorites"
) -> Dict[str, Any]:
    """
    Add a book to user's favorites list.
    
    Args:
        user_id: UUID of the user
        book_id: UUID of the book
        list_name: List type (favorites, reading, completed, plan_to_read, dropped)
        
    Returns:
        The created favorite entry
    """
    client = get_client()
    
    response = client.table("favorites").insert({
        "user_id": user_id,
        "book_id": book_id,
        "list_name": list_name
    }).execute()
    
    return response.data[0] if response.data else None


def get_user_favorites(
    user_id: str,
    list_name: Optional[str] = None
) -> List[Dict[str, Any]]:
    """
    Get user's favorite books.
    
    Args:
        user_id: UUID of the user
        list_name: Optional filter by list type
        
    Returns:
        List of favorites with book details
    """
    client = get_client()
    
    query = client.table("favorites").select(
        "*",
        "books(id, title, cover_image_url, type, status, average_rating)"
    ).eq("user_id", user_id)
    
    if list_name:
        query = query.eq("list_name", list_name)
    
    response = query.order("created_at", desc=True).execute()
    
    return response.data


# ============================================================================
# File Storage Example
# ============================================================================

def upload_cover_image(
    book_id: str,
    file_data: bytes,
    file_name: str,
    content_type: str = "image/jpeg"
) -> str:
    """
    Upload a book cover image to Supabase Storage.
    
    Args:
        book_id: UUID of the book (used in the file path)
        file_data: Image file content as bytes
        file_name: Original filename
        content_type: MIME type of the file
        
    Returns:
        Public URL of the uploaded image
    """
    client = get_admin_client()  # Using admin client for storage operations
    
    # Create a unique file path
    file_extension = file_name.split(".")[-1] if "." in file_name else "jpg"
    storage_path = f"book-covers/{book_id}.{file_extension}"
    
    # Upload to the 'covers' bucket
    response = client.storage.from_("covers").upload(
        storage_path,
        file_data,
        {"content-type": content_type}
    )
    
    # Get public URL
    public_url = client.storage.from_("covers").get_public_url(storage_path)
    
    return public_url


# ============================================================================
# Example Usage
# ============================================================================

if __name__ == "__main__":
    # Test the client
    try:
        client = get_client()
        print("✓ Supabase client initialized successfully")
        
        # Test a simple query
        # response = client.table("books").select("count", count="exact").execute()
        # print(f"✓ Found {response.count} books in database")
        
    except Exception as e:
        print(f"✗ Error: {e}")
